hail/src/main/scala/is/hail/expr/ir/InferType.scala

```diff
@@ -0,0 +1,132 @@
+package is.hail.expr.ir
+
+import is.hail.expr.types.virtual._
+import is.hail.utils._
+
+// FIXME: strip all requiredness logic when possible
+object InferType {
+  def apply(ir: IR): Type = {
+    ir match {
+      case I32(_) => TInt32()
+      case I64(_) => TInt64()
+      case F32(_) => TFloat32()
+      case F64(_) => TFloat64()
+      case Str(_) => TString()
+      case Literal(t, _) => t
+      case True() | False() => TBoolean()
+      case Void() => TVoid
+      case Cast(_, t) => t
+      case NA(t) => t
+      case IsNA(_) => TBoolean()
+      case Ref(_, t) => t
+      case In(_, t) => t
+      case MakeArray(_, t) => t
+      case _: ArrayLen => TInt32()
+      case _: ArrayRange => TArray(TInt32())
+      case _: LowerBoundOnOrderedCollection => TInt32()
+      case _: ArrayFor => TVoid
+      case _: InitOp => TVoid
+      case _: SeqOp => TVoid
+      case _: Begin => TVoid
+      case _: StringLength => TInt32()
+      case _: StringSlice => TString()
+      case Die(_, t) => t
+      case If(cond, cnsq, altr) =>
+        assert(cond.typ.isOfType(TBoolean()))
+        assert(cnsq.typ.isOfType(altr.typ))
+        if (cnsq.typ != altr.typ)
+          cnsq.typ.deepOptional()
+        else
+          cnsq.typ
+      case Let(name, value, body) =>
+        body.typ
+      case ApplyBinaryPrimOp(op, l, r) =>
+        BinaryOp.getReturnType(op, l.typ, r.typ).setRequired(l.typ.required && r.typ.required)
+      case ApplyUnaryPrimOp(op, v) =>
+        UnaryOp.getReturnType(op, v.typ).setRequired(v.typ.required)
+      case ApplyComparisonOp(op, l, r) =>
+        assert(l.typ isOfType r.typ)
+        TBoolean(l.typ.required && r.typ.required)
```

Comment:
oops---sorry to make more work for you. Can you rebase and update to include the `Compare()` case?